src/transformers/utils/chat_template_utils.py

```diff
@@ -0,0 +1,259 @@
+            },
+            "required": ["x", "y"]
+        }
+    }
+    ```
+
+    The general use for these schemas is that they are used to generate tool descriptions for chat templates that
+    support them, like so:
+
+    ```python
+    >>> from transformers import AutoTokenizer
+    >>> from transformers.utils import get_json_schema
+    >>>
+    >>> def multiply(x: float, y: float):
+    >>>    '''
+    >>>    A function that multiplies two numbers
+    >>>
+    >>>    Args:
+    >>>        x: The first number to multiply
+    >>>        y: The second number to multiply
+    >>>    return x * y
+    >>>    '''
+    >>>
+    >>> multiply_schema = get_json_schema(multiply)
+    >>> tokenizer = AutoTokenizer.from_pretrained("CohereForAI/c4ai-command-r-v01")
+    >>> messages = [{"role": "user", "content": "What is 179 x 4571?"}]
+    >>> formatted_chat = tokenizer.apply_chat_template(
+    >>>                       messages,
+    >>>                       tools=[multiply_schema],
+    >>>                       chat_template="tool_use",
+    >>>                       return_dict=True,
+    >>>                       return_tensors="pt",
+    >>>                       add_generation_prompt=True
+    >>>                       )
+    >>> # The formatted chat can now be passed to model.generate()
+    ```
+
+    In many cases, it is more convenient to simply pass the functions directly to apply_chat_template and let it
+    autogenerate schemas than calling this function directly.
+    """
+    doc = inspect.getdoc(func)
+    if not doc:
+        raise ValueError(f"Cannot generate JSON schema for {func.__name__} because it has no docstring!")
+    doc = doc.strip()
+    main_doc, param_descriptions, return_doc = parse_google_format_docstring(doc)
+
+    json_schema = _convert_type_hints_to_json_schema(func)
+    if (return_dict := json_schema["properties"].pop("return", None)) is not None:
+        if return_doc is not None:  # We allow a missing return docstring since most templates ignore it
+            return_dict["description"] = return_doc
+    for arg in json_schema["properties"]:
+        if arg not in param_descriptions:
+            raise ValueError(
+                f"Cannot generate JSON schema for {func.__name__} because the docstring has no description for the argument '{arg}'"
+            )
+        desc = param_descriptions[arg]
+        enum_choices = re.search(r"\(choices:\s*([^)]+)\)\s*$", desc, flags=re.IGNORECASE)
+        if enum_choices:
+            json_schema["properties"][arg]["enum"] = [c.strip() for c in json.loads(enum_choices.group(1))]
+            desc = enum_choices.string[: enum_choices.start()].strip()
+        json_schema["properties"][arg]["description"] = desc
+
+    output = {"name": func.__name__, "description": main_doc, "parameters": json_schema}
+    if return_dict is not None:
+        output["return"] = return_dict
+    return output
+
+
+def parse_google_format_docstring(docstring):
```

Comment:
Tricky, unfortunately, because for functions without arguments, just about any string is a valid Google-format docstring!

The best I could do is what's already in the PR - if any arguments are present in the function, and we can't parse their descriptions out of the `Args:` block in the docstring, we throw an error.